packages/components/src/tools-panel/stories/tools-panel-with-item-group-slot.js

```diff
@@ -0,0 +1,246 @@
+	const { text, background, link } = attributes;
+
+	const cx = useCx();
+	const slotWrapperClassName = cx( SlotWrapper );
+	const itemClassName = cx( ToolsPanelItemClass );
+
+	const resetAll = ( resetFilters = [] ) => {
+		let newAttributes = {};
+
+		resetFilters.forEach( ( resetFilter ) => {
+			newAttributes = {
+				...newAttributes,
+				...resetFilter( newAttributes ),
+			};
+		} );
+
+		setAttributes( newAttributes );
+	};
+
+	const updateAttribute = ( name, value ) => {
+		setAttributes( {
+			...attributes,
+			[ name ]: value,
+		} );
+	};
+
+	const ToolsPanelColorDropdown = ( { attribute, label, value } ) => {
+		return (
+			<ToolsPanelItem
+				className={ itemClassName }
+				hasValue={ () => !! value }
+				label={ label }
+				onDeselect={ () => updateAttribute( attribute, undefined ) }
+				resetAllFilter={ () => ( { [ attribute ]: undefined } ) }
+				panelId={ panelId }
+				as={ Item }
+			>
+				<Dropdown
+					renderToggle={ ( { onToggle } ) => (
+						<Button onClick={ onToggle }>
+							<HStack justify="flex-start">
+								<ColorIndicator colorValue={ value } />
+								<FlexItem>{ label }</FlexItem>
+							</HStack>
+						</Button>
+					) }
+					renderContent={ () => (
+						<ColorPalette
+							value={ value }
+							colors={ colors }
+							onChange={ ( newColor ) =>
+								updateAttribute( attribute, newColor )
+							}
+						/>
+					) }
+				/>
+			</ToolsPanelItem>
+		);
+	};
+
+	// ToolsPanelItems are rendered via two different fills to simulate
+	// injection from multiple locations.
+	return (
+		<SlotFillProvider>
+			<PanelWrapperView>
+				<Panel>
+					<ToolsPanelItems.Slot
+						as={ ItemGroup }
+						className={ slotWrapperClassName }
+						resetAll={ resetAll }
+					/>
+				</Panel>
+			</PanelWrapperView>
+			<ToolsPanelItems>
+				<ToolsPanelColorDropdown
+					attribute="text"
+					label="Text"
+					value={ text }
+				/>
+				<ToolsPanelColorDropdown
+					attribute="background"
+					label="Background"
+					value={ background }
+				/>
+			</ToolsPanelItems>
+			<ToolsPanelItems>
+				<ToolsPanelColorDropdown
+					attribute="link"
+					label="Link"
+					value={ link }
+				/>
+			</ToolsPanelItems>
+		</SlotFillProvider>
+	);
+};
+
+const PanelWrapperView = styled.div`
+	max-width: 280px;
+	font-size: 13px;
+
+	.components-dropdown-menu__menu {
+		max-width: 220px;
+	}
+`;
+
+const SlotWrapper = css`
+	&&& {
+		row-gap: 0;
+	}
+
+	> div {
+		grid-column: span 2;
+	}
+`;
+
+const ToolsPanelItemClass = css`
+	padding: 0;
+	border-left: 1px solid rgba( 0, 0, 0, 0.1 );
+	border-right: 1px solid rgba( 0, 0, 0, 0.1 );
+	border-bottom: 1px solid rgba( 0, 0, 0, 0.1 );
+
+	&&.first {
+		border-top-left-radius: 10px;
+		border-top-right-radius: 10px;
+		border-top: 1px solid rgba( 0, 0, 0, 0.1 );
+	}
+
+	&.last {
+		border-bottom-left-radius: 10px;
+		border-bottom-right-radius: 10px;
+	}
+
+	&& > div,
+	&& > div > button {
+		width: 100%;
+		border-radius: inherit;
+	}
```

Comment:
By using `ItemGroup`'s `isBordered` and `isSeparated` props, we don't need to define custom CSS for borders (apart from a one-line override for the `last` item)

<details>
<summary>see diff</summary>

```diff
diff --git a/packages/components/src/tools-panel/stories/tools-panel-with-item-group-slot.js b/packages/components/src/tools-panel/stories/tools-panel-with-item-group-slot.js
index 950a3c3f5e..ae4eaad17a 100644
--- a/packages/components/src/tools-panel/stories/tools-panel-with-item-group-slot.js
+++ b/packages/components/src/tools-panel/stories/tools-panel-with-item-group-slot.js
@@ -169,6 +169,8 @@ export const ToolsPanelWithItemGroupSlot = () => {
 				<Panel>
 					<ToolsPanelItems.Slot
 						as={ ItemGroup }
+						isBordered
+						isSeparated
 						className={ slotWrapperClassName }
 						resetAll={ resetAll }
 					/>
@@ -218,29 +220,18 @@ const SlotWrapper = css`
 
 const ToolsPanelItemClass = css`
 	padding: 0;
-	border-left: 1px solid rgba( 0, 0, 0, 0.1 );
-	border-right: 1px solid rgba( 0, 0, 0, 0.1 );
-	border-bottom: 1px solid rgba( 0, 0, 0, 0.1 );
-
-	&&.first {
-		border-top-left-radius: 10px;
-		border-top-right-radius: 10px;
-		border-top: 1px solid rgba( 0, 0, 0, 0.1 );
+
+	&.first {
+		border-left: 2px solid blue;
 	}
 
 	&.last {
-		border-bottom-left-radius: 10px;
-		border-bottom-right-radius: 10px;
+		border-right: 2px solid red;
+		border-bottom-color: transparent;
 	}
 
 	&& > div,
 	&& > div > button {
 		width: 100%;
-		border-radius: inherit;
-	}
-
-	/* .components-dropdown class overrides ToolsPanelItemPlaceholder styles */
-	&[class*='ToolsPanelItemPlaceholder'] {
-		display: none;
 	}
 `;
```
</details>

Reply:
Thanks. These were leftover from when the `Item` wasn't a direct child of the slot. I'll clean it up.

Reply:
Revisiting this, the proposed suggestion won't really work for our needs and jogged my memory on why the original approach was taken. 

There are a couple of issues:
- The `border-radius: inherit` for the button is needed so the focus outline is consistent with the rounded corners
- We still need the application of the `border-radius` on the `.first` and `.last` classes because the `ItemGroup` styling doesn't handle hidden elements. (Part of the whole reason for this PR in the first place)
- Using `isBordered` and `isSeparated` forces everything to visually be represented as an item. 

Regarding the last point, it is a problem when in our real-world use case, a contrast checker is also rendered into the panel. This means the contrast checker is not a `ToolsPanelItem` but appears within the slot which is technically rendered as an `ItemGroup`. Not forcing `isBordered` and `isSeparated` means it can be visually separated.

<details>
<summary>GIFs demostrating `isBordered` issue</summary>

| Suggested Changes | Current Approach |
|---|---|
| ![ToolsPanelFirstLastContrastChecker](https://user-images.githubusercontent.com/60436221/149439537-70cc33b0-305e-43cc-8bdb-72468a3d3aad.gif) | ![ToolsPanelFirstLastContrastChecker2](https://user-images.githubusercontent.com/60436221/149439562-83b01dcc-6bf8-4996-9c1f-adb83cbd3e4e.gif) |
</details>

I've updated the story to include a contrast checker.

Reply:
> Could you show a scenario in which ItemGroup would fail in this scenario? When I came up with the code path above, I tested the Storybook example and everything seems to work fine when showing/hiding toolspanel items.

The inheritance of the border-radius doesn't work fine because the `ItemGroup` styles rely on `> *:first-of-type > *` and `> *:last-of-type > *` to apply the `rounded` style border radii. These styles can't take into account whether the element in the item is actually just a non-visible placeholder.

I've updated the story to explicitly disable `isRounded` on the slot's `ItemGroup`. As noted previously, we still need styles to handle the border-radius on the `.first` and `.last` items. They can be made to inherit from the wrapping `ItemGroup` (with an extra rule to target the item wrapper), exactly how this styling happens isn't really the issue. We could apply a custom border-radius to the `ItemGroup` then inherit on the appropriate items or just style those first/last items directly. Either way these first/last classes were needed.

With all the iterations on this, I didn't remove the border color and styling part of the styles as I was distracted by the contrast checker issue when looking at the implications for https://github.com/WordPress/gutenberg/pull/34027.

> There's an issue, though: ItemGroup would technically expect ContrastChecker to be rendered inside an Item.

This is a fair point and is relevant to https://github.com/WordPress/gutenberg/pull/34027. Either way, we still need the classes this PR provides and I don't think should be a blocker here.

I'll remove the contrast checker from the example here so as to avoid future confusion.

> A potential solution could be to render ItemGroup as a child of ToolsPanel:

With a view towards https://github.com/WordPress/gutenberg/pull/34027, this is easier said than done. Any block or plugin could add new color options to be included within the Inspector Controls color panel. This means the fills that render in the slot come from different places and may contain a mix of color options, contrast checkers or any other element they wish to add into the panel (help text etc).

Moving the contrast checkers to the bottom of the panel was done via CSS ordering in https://github.com/WordPress/gutenberg/pull/34027. 

I don't believe we can programmatically control the fills, especially when we don't know exactly what we'll be dealing with, so to make the `ItemGroup` color options only and an inner element of the `ToolsPanel` we'd likely need to create one or two new slots and nest those inside an already nested slot in the block editor. It makes the color panel a special case with the Inspector Controls group slots and probably more difficult for consumers.

Given that, I'd be inclined in https://github.com/WordPress/gutenberg/pull/34027 to no longer use the `ItemGroup` at all.

> If that didn't work, maybe ItemGroup is not the right fit for this scenario

Agreed. This is more a hangover of the color panel being updated without consideration or knowledge that there was a PR open for review to move it to using the `ToolsPanel`.

I'll update https://github.com/WordPress/gutenberg/pull/34027 to remove the use of `ItemGroup` for the colors panel. 

That is obviously separate to this PR so I think this should be ok to land. Since the decision to adopt the new first/last classname props was accepted we've only been iterating on the storybook example.

Reply:
Sorry for the delayed response (has a couple of setbacks this week)!

I agree with all of your points — the style rules (and semantics requirements) related to `ItemGroup` are very strict, and everything is made even more complicated by the fact that tool panel items can be added via slot-fills.

In the light of our findings I think that the best way forward, at least for now, is to **avoid using `ItemGroup` in combination with `ToolsPanel`** — both in #34027 and in this PR (you may even, at this point, re-introduce the contrast checker).

We will also have a look at how to improve the `ItemGroup` component (of course separately from this PR), in order to make it more flexible/adaptable to different situations (both in terms of styles and in terms of the list semantics).